File: projects/src/main/java/io/nirahtech/erp/projects/Task.java
package io.nirahtech.erp.projects;

import java.io.File;
import java.io.Serializable;
import java.time.Duration;
import java.time.LocalDate;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

public final class Task implements Serializable {

    private final TaskIdentifier id;
    private final String name;
    private final Task parent;
    private String details;
    private Status status;
    private LocalDate startDate;
    private Duration duration;
    private Priority prority;
    private ProjectMember referent;
    private int maxWorkers;
    private final Set<Task> subTasks = new HashSet();
    private final Set<Task> requiredTasks = new HashSet<>();
    private final Set<File> documentations = new HashSet<>();
    private final Set<String> labels = new HashSet<>();
    private final Set<ProjectMember> workers = new HashSet<>();
    private final Map<Status, Runnable> eventListeners = new HashMap<>();

    public Task(
            TaskIdentifier id,
            String name,
            String details,
            Status status,
            Task parent,
            Set<Task> subTasks,
            Set<Task> requiredTasks,
            LocalDate startDate,
            Duration duration,
            Priority prority,
            Set<String> labels,
            Set<File> documentations,
            ProjectMember referent,
            Set<ProjectMember> workers,
            int maxWorkers) {
        this.id = id;
        this.name = name;
        this.details = details;
        this.status = status;
        this.parent = parent;
        this.subTasks.addAll(subTasks);
        this.requiredTasks.addAll(requiredTasks);
        this.startDate = startDate;
        this.duration = duration;
        this.prority = prority;
        this.labels.addAll(labels);
        this.documentations.addAll(documentations);
        this.referent = referent;
        this.workers.addAll(workers);
        this.maxWorkers = maxWorkers;
    }

    public String getDetails() {
        return details;
    }

    public Set<File> getDocumentations() {
        return documentations;
    }

    public Duration getDuration() {
        return duration;
    }

    public TaskIdentifier getId() {
        return id;
    }

    public Set<String> getLabels() {
        return labels;
    }

    public int getMaxWorkers() {
        return maxWorkers;
    }

    public String getName() {
        return name;
    }

    public Task getParent() {
        return parent;
    }

    public Priority getPrority() {
        return prority;
    }

    public ProjectMember getReferent() {
        return referent;
    }

    public Set<Task> getRequiredTasks() {
        return requiredTasks;
    }

    public LocalDate getStartDate() {
        return startDate;
    }

    public Status getStatus() {
        return status;
    }

    public Set<Task> getSubTasks() {
        return subTasks;
    }

    public Set<ProjectMember> getWorkers() {
        return workers;
    }

    public void setDetails(String details) {
        this.details = details;
    }

    public void setDuration(Duration duration) {
        this.duration = duration;
    }

    public void setMaxWorkers(int maxWorkers) {
        this.maxWorkers = maxWorkers;
    }

    public void setPrority(Priority prority) {
        this.prority = prority;
    }

    public void setReferent(ProjectMember referent) {
        this.referent = referent;
    }

    public void setStartDate(LocalDate startDate) {
        this.startDate = startDate;
    }

    public void setStatus(Status status) {
		this.status = status;
		if (this.eventListeners.containsKey(status)) {
			Runnable callback = this.eventListeners.get(status);
			callback.run();
		}
    }

    public final LocalDate getDeadLine() {
        return this.startDate.plusDays(this.duration.toDays());
    }

    public void addEventListenerOnStatusChanged(final Status status, final Runnable callback) {
        this.eventListeners.put(status, callback);
    }
}

File: webapp/src/main/react/src/data/Task.ts
import { File } from "./File";
import { ProjectMember } from "./ProjectMember";
import { Status } from "./Status";
import { TaskIdentifier } from "./TaskIdentifier";

class Task {
  private readonly id: TaskIdentifier;
  private readonly name: string;
  private readonly parent: Task | null;
  private details: string;
  private status: Status;
  private startDate: Date;
  private duration: Duration;
  private prority: Priority;
  private referent: ProjectMember;
  private maxWorkers: number;
  private readonly subTasks: Set<Task>;
  private readonly requiredTasks: Set<Task>;
  private readonly documentations: Set<File>;
  private readonly labels: Set<string>;
  private readonly workers: Set<ProjectMember>;
  private readonly eventListeners: Map<Status, Runnable>;

  constructor(
    id: TaskIdentifier,
    name: string,
    details: string,
    status: Status,
    parent: Task | null,
    subTasks: Set<Task>,
    requiredTasks: Set<Task>,
    startDate: Date,
    duration: Duration,
    prority: Priority,
    labels: Set<string>,
    documentations: Set<File>,
    referent: ProjectMember,
    workers: Set<ProjectMember>,
    maxWorkers: number
  ) {
    this.id = id;
    this.name = name;
    this.details = details;
    this.status = status;
    this.parent = parent;
    this.subTasks = new Set<Task>(subTasks);
    this.requiredTasks = new Set<Task>(requiredTasks);
    this.startDate = startDate;
    this.duration = duration;
    this.prority = prority;
    this.labels = new Set<string>(labels);
    this.documentations = new Set<File>(documentations);
    this.referent = referent;
    this.workers = new Set<ProjectMember>(workers);
    this.maxWorkers = maxWorkers;
    this.eventListeners = new Map<Status, Runnable>();
  }

  getDetails(): string {
    return this.details;
  }

  getDocumentations(): Set<File> {
    return this.documentations;
  }

  getDuration(): Duration {
    return this.duration;
  }

  getId(): TaskIdentifier {
    return this.id;
  }

  getLabels(): Set<string> {
    return this.labels;
  }

  getMaxWorkers(): number {
    return this.maxWorkers;
  }

  getName(): string {
    return this.name;
  }

  getParent(): Task | null {
    return this.parent;
  }

  getPrority(): Priority {
    return this.prority;
  }

  getReferent(): ProjectMember {
    return this.referent;
  }

  getRequiredTasks(): Set<Task> {
    return this.requiredTasks;
  }

  getStartDate(): Date {
    return this.startDate;
  }

  getStatus(): Status {
    return this.status;
  }

  getSubTasks(): Set<Task> {
    return this.subTasks;
  }

  getWorkers(): Set<ProjectMember> {
    return this.workers;
  }

  setDetails(details: string): void {
    this.details = details;
  }

  setDuration(duration: Duration): void {
    this.duration = duration;
  }

  setMaxWorkers(maxWorkers: number): void {
    this.maxWorkers = maxWorkers;
  }

  setPrority(prority: Priority): void {
    this.prority = prority;
  }

  setReferent(referent: ProjectMember): void {
    this.referent = referent;
  }

  setStartDate(startDate: Date): void {
    this.startDate = startDate;
  }

  setStatus(status: Status): void {
    this.status = status;
    if (this.eventListeners.has(status)) {
      const callback = this.eventListeners.get(status);
      if (callback) {
        callback();
      }
    }
  }

  getDeadLine(): Date {
    const endDate = new Date(this.startDate);
    endDate.setDate(endDate.getDate() + this.duration);
    return endDate;
  }

  addEventListenerOnStatusChanged(status: Status, callback: Runnable): void {
    this.eventListeners.set(status, callback);
  }
}

export { Task };
